File: MiniGame-Java/src/main/java/rs/lab/mges/game/GameAssets.java
package rs.lab.mges.game;

import io.github.libsdl4j.api.render.SDL_Renderer;
import io.github.libsdl4j.api.render.SDL_Texture;
import java.nio.file.Path;
import rs.lab.mges.engine.AssetsLoader;
import rs.lab.mges.engine.Sound;

import static io.github.libsdl4j.api.render.SdlRender.SDL_DestroyTexture;

public final class GameAssets {

    public SDL_Texture[] playerFrames;
    public SDL_Texture[] beeFrames;
    public SDL_Texture[] blueFrames;
    public SDL_Texture[] snikyFrames;
    public SDL_Texture[] foxyFrames;
    public SDL_Texture[] heroBFrames;

    public SDL_Texture playerBullet;
    public SDL_Texture ground;
    public SDL_Texture bloodGround;
    public SDL_Texture bloodParticle;
    public SDL_Texture explosion;
    public SDL_Texture healthbar;
    public SDL_Texture elements;
    public SDL_Texture title;
    public SDL_Texture medKit;
    public SDL_Texture forest;
    public SDL_Texture bomb;

    public Sound.Audio music;
    public Sound.Audio tickFx;
    public Sound.Audio explosionFx;
    public Sound.Audio pickupFx;
    public Sound.Audio mobHitFx;
    public Sound.Audio fireFx;
    public Sound.Audio explosionLongFx;

    private final SDL_Renderer renderer;

    public GameAssets(SDL_Renderer renderer) {
        this.renderer = renderer;

        Sound.initAudio();

        playerFrames = new SDL_Texture[]{
            loadFrame("hero1.bmp"),
            loadFrame("hero2.bmp")
        };

        beeFrames = new SDL_Texture[]{
            loadFrame("bee1a.bmp"),
            loadFrame("bee2a.bmp"),};

        blueFrames = new SDL_Texture[]{
            loadFrame("blue1a.bmp"),
            loadFrame("blue2a.bmp"),};

        snikyFrames = new SDL_Texture[]{
            loadFrame("sniky1.bmp"),
            loadFrame("sniky2.bmp"),};

        foxyFrames = new SDL_Texture[]{
            loadFrame("roka1.bmp"),
            loadFrame("roka2.bmp"),};

        heroBFrames = new SDL_Texture[]{
            loadFrame("hero1b.bmp"),
            loadFrame("hero2b.bmp"),};

        music = loadSound("world_1.wav");

        elements = AssetsLoader.loadBitmap(renderer, "Bitmaps/elements.bmp");
        title = AssetsLoader.loadBitmap(renderer, "Bitmaps/title.bmp");
        forest = AssetsLoader.loadBitmap(renderer, "Bitmaps/background.bmp");

        playerBullet = loadFrame("bullet.bmp");
        ground = loadFrame("tile_0004.bmp");

        bloodGround = loadFrame("blood_tile.bmp");
        bloodParticle = loadFrame("blood.bmp");

        explosion = loadFrame("explosion-4.bmp");
        healthbar = loadFrame("healthbar.bmp");

        medKit = loadFrame("medkit.bmp");
        bomb = loadFrame("bombball.bmp");
        tickFx = loadSound("tick.wav");
        explosionFx = loadSound("explosion.wav");
        pickupFx = loadSound("pickup.wav");
        mobHitFx = loadSound("hit.wav");
        fireFx = loadSound("fire.wav");
        explosionLongFx = loadSound("explosion_long.wav");
    }

    public void freeAll() {
        SDL_DestroyTexture(playerFrames[0]);
        SDL_DestroyTexture(playerFrames[1]);
        SDL_DestroyTexture(beeFrames[0]);
        SDL_DestroyTexture(beeFrames[1]);
        SDL_DestroyTexture(blueFrames[0]);
        SDL_DestroyTexture(blueFrames[1]);
        SDL_DestroyTexture(snikyFrames[0]);
        SDL_DestroyTexture(snikyFrames[1]);
        SDL_DestroyTexture(foxyFrames[0]);
        SDL_DestroyTexture(foxyFrames[1]);

        SDL_DestroyTexture(playerBullet);

        SDL_DestroyTexture(ground);
        SDL_DestroyTexture(bloodGround);
        SDL_DestroyTexture(bloodParticle);
        SDL_DestroyTexture(explosion);
        SDL_DestroyTexture(healthbar);
        SDL_DestroyTexture(title);
        SDL_DestroyTexture(elements);

        Sound.freeAudio(music);
        Sound.freeAudio(tickFx);
        Sound.freeAudio(explosionFx);
        Sound.freeAudio(pickupFx);
        Sound.freeAudio(mobHitFx);
        Sound.freeAudio(fireFx);
        Sound.freeAudio(explosionLongFx);

        Sound.endAudio();
    }

    private SDL_Texture loadFrame(String fileName) {
        return AssetsLoader.loadBitmap(renderer, Path.of("Tiles", fileName).toString());
    }

    private Sound.Audio loadSound(String fileName) {
        return AssetsLoader.loadSoundFx(Path.of("Sounds", fileName).toString());
    }
}

File: MiniGame-NET/MiniGame/GameAssets.cs
// Copyright 2023 Peter Bakota
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using MiniGameEngine;
using SDL2;

namespace MiniGame;

class GameAssets: AssetsLoader
{
    public IntPtr[] PlayerFrames;
    public IntPtr[] BeeFrames;
    public IntPtr[] BlueFrames;
    public IntPtr[] SnikyFrames;
    public IntPtr[] FoxyFrames;
    public IntPtr[] HeroBFrames;
    public IntPtr PlayerBullet;
    public IntPtr Ground;
    public IntPtr BloodGround;
    public IntPtr BloodParticle;
    public IntPtr Explosion;
    public IntPtr Healthbar;
    public IntPtr Elements;
    public IntPtr Title;
    public IntPtr MedKit;
    public IntPtr Forest;
    public IntPtr Bomb;
    public Sound.Audio Music;
    public Sound.Audio TickFx;
    public Sound.Audio ExplosionFx;
    public Sound.Audio PickupFx;
    public Sound.Audio MobHitFx;
    public Sound.Audio FireFx;
    public Sound.Audio ExplosionLongFx;

    public GameAssets(IntPtr renderer):base("assets")
    {
        PlayerFrames = new IntPtr[] {
            LoadFrame(renderer, "hero1.png"),
            LoadFrame(renderer, "hero2.png"),
        };

        BeeFrames = new IntPtr[] {
            LoadFrame(renderer, "bee1a.png"),
            LoadFrame(renderer, "bee2a.png"),
        };

        BlueFrames = new IntPtr[] {
            LoadFrame(renderer, "blue1a.png"),
            LoadFrame(renderer, "blue2a.png"),
        };

        SnikyFrames = new IntPtr[] {
            LoadFrame(renderer, "sniky1.png"),
            LoadFrame(renderer, "sniky2.png"),
        };

        FoxyFrames = new IntPtr[] {
            LoadFrame(renderer, "roka1.png"),
            LoadFrame(renderer, "roka2.png"),
        };

        HeroBFrames = new IntPtr[] {
            LoadFrame(renderer, "hero1b.png"),
            LoadFrame(renderer, "hero2b.png"),
        };

        Music = LoadSound("world_1.wav");

        PlayerBullet = LoadFrame(renderer, "bullet.png");
        Ground = LoadFrame(renderer, "tile_0004.png");

        BloodGround = LoadFrame(renderer, "blood_tile.png");
        BloodParticle = LoadFrame(renderer, "blood.png");

        Explosion = LoadFrame(renderer, "explosion-4.png");
        Healthbar = LoadFrame(renderer, "healthbar.png");

        Elements = LoadBitmap(renderer, "Bitmaps/elements.png");
        Title = LoadBitmap(renderer, "Bitmaps/title.png");
        MedKit = LoadFrame(renderer, "medkit.png");
        Forest = LoadBitmap(renderer, "Bitmaps/background.png");

        Bomb = LoadFrame(renderer, "bombball.png");
        TickFx = LoadSound("tick.wav");
        ExplosionFx = LoadSound("explosion.wav");
        PickupFx = LoadSound("pickup.wav");
        MobHitFx = LoadSound("hit.wav");
        FireFx = LoadSound("fire.wav");
        ExplosionLongFx = LoadSound("explosion_long.wav");
    }

    public void FreeAll()
    {
        Sound.FreeAudio(Music);
        Sound.FreeAudio(TickFx);
        Sound.FreeAudio(ExplosionFx);
        Sound.FreeAudio(PickupFx);
        Sound.FreeAudio(MobHitFx);
        Sound.FreeAudio(FireFx);
        Sound.FreeAudio(ExplosionLongFx);

        SDL.SDL_DestroyTexture(PlayerFrames[0]);
        SDL.SDL_DestroyTexture(PlayerFrames[1]);
        SDL.SDL_DestroyTexture(BeeFrames[0]);
        SDL.SDL_DestroyTexture(BeeFrames[1]);
        SDL.SDL_DestroyTexture(BlueFrames[0]);
        SDL.SDL_DestroyTexture(BlueFrames[1]);
        SDL.SDL_DestroyTexture(SnikyFrames[0]);
        SDL.SDL_DestroyTexture(SnikyFrames[1]);
        SDL.SDL_DestroyTexture(FoxyFrames[0]);
        SDL.SDL_DestroyTexture(FoxyFrames[1]);

        SDL.SDL_DestroyTexture(PlayerBullet);

        SDL.SDL_DestroyTexture(Ground);
        SDL.SDL_DestroyTexture(BloodGround);
        SDL.SDL_DestroyTexture(BloodParticle);
        SDL.SDL_DestroyTexture(Explosion);
        SDL.SDL_DestroyTexture(Healthbar);
        SDL.SDL_DestroyTexture(Title);
        SDL.SDL_DestroyTexture(Elements);
    }

    private static IntPtr LoadFrame(IntPtr renderer, string fileName)
        => LoadBitmap(renderer, Path.Combine("Tiles", fileName));

    private static Sound.Audio LoadSound(string fileName)
        => LoadSoundFx(Path.Combine("Sounds", fileName));
}